fuse_core/src/async_motion_model.cpp

```diff
@@ -62,56 +62,77 @@ bool AsyncMotionModel::apply(Transaction& transaction)
   // Thus, it is functionally similar to a service callback, and should be a familiar pattern for ROS developers.
   // This function blocks until the queryCallback() call completes, thus enforcing that motion models are generated
   // in order.
-  auto callback = boost::make_shared<CallbackWrapper<bool>>(
+  auto callback = std::make_shared<CallbackWrapper<bool>>(
     std::bind(&AsyncMotionModel::applyCallback, this, std::ref(transaction)));
   auto result = callback->getFuture();
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
+  callback_queue_->addCallback(callback);
   result.wait();
-  return result.get();
-}
 
-void AsyncMotionModel::graphCallback(Graph::ConstSharedPtr graph)
-{
-  callback_queue_.addCallback(
-    boost::make_shared<CallbackWrapper<void>>(std::bind(&AsyncMotionModel::onGraphUpdate, this, std::move(graph))),
-    reinterpret_cast<uint64_t>(this));
+  return result.get();
 }
 
 void AsyncMotionModel::initialize(const std::string& name)
 {
   // Initialize internal state
   name_ = name;
-  node_handle_.setCallbackQueue(&callback_queue_);
-  private_node_handle_ = ros::NodeHandle("~/" + name_);
-  private_node_handle_.setCallbackQueue(&callback_queue_);
+  std::string node_namespace = "";
+
+  rclcpp::Context::SharedPtr ros_context = rclcpp::contexts::get_global_default_context();
+  auto node_options = rclcpp::NodeOptions();
+  node_options.context(ros_context); //set a context to generate the node in
+
+  node_ = rclcpp::Node::make_shared(name_, node_namespace, node_options);
+
+  auto executor_options = rclcpp::ExecutorOptions();
+  executor_options.context = ros_context;
+  executor_ = rclcpp::executors::MultiThreadedExecutor::make_shared(executor_options, executor_thread_count_);
+
+  callback_queue_ = std::make_shared<CallbackAdapter>(ros_context);
+  node_->get_node_waitables_interface()->add_waitable(
+    callback_queue_, (rclcpp::CallbackGroup::SharedPtr) nullptr);
 
   // Call the derived onInit() function to perform implementation-specific initialization
   onInit();
 
-  // Start the async spinner to service the local callback queue
-  spinner_.start();
+  executor_->add_node(node_);
+}
+
+void AsyncMotionModel::graphCallback(Graph::ConstSharedPtr graph)
+{
+  auto callback = std::make_shared<CallbackWrapper<void>>(
+    std::bind(&AsyncMotionModel::onGraphUpdate, this, std::move(graph))
+  );
+  auto result = callback->getFuture();
+  callback_queue_->addCallback(callback);
+  result.wait();
```

Comment:
I think the intent of the separate queue and executor is to not wait for the result here.

Reply:
I removed it everywhere where it made sense to https://github.com/locusrobotics/fuse/pull/284/commits/d3b6c944272008f0ad4f62582c73deb350f3b9e1

I commented them out and added a TODO though, just in case we need to wait due to race conditions/etc. when everything compiles

Reply:
The futures and waits are all about thread management.

The AsyncMotionModel is running its own callback spinner thread. So if a derived implementation created a Timer or subscribed to a sensor topic, those callbacks will run in the AsyncMotionModel spinner thread. However, the `apply()` method here is going to be called by the Optimizer, and hence it will run in the Optimizer's thread. This means any derived motion model must be very careful about thread synchronization. And it is not very obvious what callbacks will run in what threads. To make it easier to write derived motion model classes, I implemented the following:

The `apply()` method will be called by the Optimizer thread
```
bool AsyncMotionModel::apply(Transaction& transaction)
{
```

Wrap the user-defined implementation, `applyCallback()`, inside a promise.
```
  auto callback = boost::make_shared<CallbackWrapper<bool>>(
    std::bind(&AsyncMotionModel::applyCallback, this, std::ref(transaction)));
```

Create a future from the wrapper so that the `apply()` function may receive the results of the `applyCallback()` function.
```
auto result = callback->getFuture();
```

Insert the `applyCallback()` call _into the callback queue of the AsyncMotionModel_. This means that the `applyCallback()` method will be executed by the AsyncMotionModel spinner thread.
```
callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
```

Wait for the AsyncMotionModel spinner thread to complete the `applyCallback()` call.
```
result.wait();
```

Return the results of the `applyCallback()` method out of the `apply()` method.
```
return result.get();
```

This was done so that the user-implementation of the `apply()` business logic, implemented in the `applyCallback()` method, will execute synchronously with any other callbacks (timers, subscriptions, services, etc) defined in the motion model. Typically the author of a derived AsyncMotionModel does not need to implement any mutexes or other thread synchronization mechanisms even though the execution of a motion model involves multiple threads created in different parts of the fuse ecosystem.